src/WinRT.Runtime/ComWrappersSupport.net5.cs

```diff
@@ -446,7 +450,14 @@ private static object CreateObject(IObjectReference objRef)
                     // we use the Inspectable wrapper directly.
                     return inspectable;
                 }
-                return ComWrappersSupport.GetTypedRcwFactory(runtimeClassName)(inspectable);
+
+                var obj = ComWrappersSupport.GetTypedRcwFactory(runtimeClassName)(inspectable);
+                var type = TypeNameSupport.FindTypeByNameCached(runtimeClassName);
+                if (type != null && (type.IsValueType || runtimeClassName.StartsWith("Windows.Foundation.IReferenceArray`1")))

```

Comment:
You can do the following instead, which should be slightly cheaper and just as accurate.

```suggestion
                if (type != null && (type.IsValueType || type.IsArray))
```

Reply:
@jkoritzinsky I just tried it, but it does not seem to work. `type.IsArray` is false even though `runtimeClassName.StartsWith("Windows.Foundation.IReferenceArray`1")` is true.

Reply:
What type does `type` represent in the non-working case? Just so I can understand what the failure condition is.

Reply:
@jkoritzinsky 
```
{Name = "IReferenceArray`1" FullName = "Windows.Foundation.IReferenceArray`1[[System.String, System.Private.CoreLib, Version=5.0.0.0, Culture=neutral, PublicKeyToken=7cec85d7bea7798e]]"}
```

Reply:
Right, it would grab the IReferenceArray interface in WinRT.Runtime. In that case, I think the only options are either a string comparison or checking if the type is generic and it's definition is `typeof(IReferenceArray<>)`. So the string comparison is probably much faster then.

Reply:
Maybe you could have `GetTypedRcwFactory` have an `out bool` parameter that specifies that an "unboxing" factory was generated, and have `GetTypedRcwFactory` set that to true whenever `GetTypedRcwFactory` calls `CreateKeyValuePairFactory`, `CreateNullableTFactory`, or `IsIReferenceArray` is called.

Since those are the exact three cases that need the special-casing for this optimization as well (and any other future cases caused by new well-known projections from WinRT types to value types would likely also need a special factory in the same spot so this would be easily extensible).

Then in this method, you would just check if the call to `GetTypedRcwFactory` set its "was an unboxing factory generated" out parameter to true to decide to put the value in the cache.

Reply:
You could even wrap the methods returned by those factories to automatically insert the value into the cache if you want to make it even cleaner and more localized (this would require making the `ConditionalWeakTable` static and live in `ComWrappersSupport`, but as long as the writing happens during a `CreateObject` method call, it will work correctly.)

Reply:
@jkoritzinsky Since the second way requires adding more reflection-based code in the `CreateX` calls, I wonder if that would be more expensive than the first method you suggest.

Reply:
For the first way, since the factory itself is cached, we would also have to cache the boolean result of whether the unboxing factory was created. So we store the cache now as `KeyValuePair<Func<>, bool>` instead if just `Func<>` - which adds memory for each type used and not just value types. I wonder if the time-memory tradeoff is worth it.

Reply:
I think the second way doesn't actually require any extra reflection-based code.

I think you could do something like the following:

```csharp

private static ConditionalWeakTable<object, IInspectable> _boxedValueReferenceCache = new();

private static Func<IInspectable, object> CreateReferenceCachingFactory(Func<IInspectable, object> internalFactory)
{
	return inspectable =>
    {
          object resultingObject = internalFactory(inspectable);
          _boxedValueReferenceCache.Add(resultingObject, inspectable);
         return resultingObject;
    };
}

```

And then change the three callers (one given as an example):

```diff
            if (implementationType.IsGenericType && implementationType.GetGenericTypeDefinition() == typeof(System.Collections.Generic.KeyValuePair<,>))
            {
-                 return CreateKeyValuePairFactory(implementationType);
+                 return CreateReferenceCachingFactory(CreateKeyValuePairFactory(implementationType));
            }
```

This way, the entire implementation of this feature is shared between the netstandard2.0 and net5.0+ implementations and still occurs at the correct time. Additionally, the special behavior ends up being cached per-type, which is a nice perf boost too (don't need to check for special behavior per-RCW, just per type).

Reply:
Thank you for this suggestion. I made this change. Can you take a look at it